mlos_bench/mlos_bench/environments/__init__.py

```diff
@@ -72,6 +102,50 @@
 `globals`, allowing for configs to be more modular and composable.
 
+Variable Propagation
+++++++++++++++++++++
+
+Parameters declared in the ``const_args`` or ``required_args`` sections of the environment config
+can be overridden with values of the corresponding parameters of the parent environment or
+specified in the external config files or the command line. In fact, ``const_args`` or
+``required_args`` sections can be viewed as placeholders for the parameters that are being pushed
+to the environment from the outside.
+
+The same parameter can be present in both ``const_args`` and ``required_args`` sections.
+``required_args`` is just a way to emphasize the importance of the parameter and create a
+placeholder for it when no default value can be specified the ``const_args`` section.
+
+Variable replacement happens in the bottom-up manner. That is, if a certain parameter is present
+in the parent (composite) environment, it will replace the corresponding parameter in the child,
+and so on. Note that the parameter _must_ appear in the child environment ``const_args`` or
+``required_args`` section; if a parameter is not present in one of these placeholders of the child
+environment config, it will not be propagated. This hierarchy allows MLOS users to have small
+immutable environment configurations at the lower levels and combine and parameterize them at the
+higher levels.
+
+Taking it to the next level outside of the environment configs, the parameters can be defined in
+the external key-value JSON config files (usually referred to as "global config files" in MLOS
+lingo). Users can have multiple global config files; at runtime, parameters from these files will
+be combined into a single dictionary and pushed to the root environment. This way, users can keep
+their experiment-specific parameters separately from the environment configs making them more
+reusable. Another common use of global config files is to store sensitive data (e.g., passwords,
+tokens, etc.) that should not be version-controlled. The global config files are specified in the
+``globals`` section of the top-level CLI config, or in the ``--globals`` command line parameter.
+
+Finally, any global or :py:class:`~.Environment` parameter can be
+overridden from the command line, by simply specifying ``--PARAMETER_NAME PARAMETER_VALUE``.
+
+We can summarize the parameter propagation rules as follows:
+
+1. Child environment will only get the parameters defined in its ``const_args`` or
+   ``required_args`` sections.
+2. Value of the parameter defined in the ``const_args`` section of the parent environment will
+   override the value of the corresponding parameter in the child environments.
+3. Values of the parameters defined in the global config files will override the values of the
+   corresponding parameters in all environments.
+4. Values of the command line parameters take precedence over values defined in the global or
+   environment configs.
+
```

Comment:
I think we should probably have an Example section that describes the flow of variables from globals, shell environments, const_args, required, run phase commands, etc. like with actual JSON blocks and commands and output.

Something super simple that folks can browse thru to get the general idea and then another link to a section of the code or tests that they can go to see more information.

Sound good?

Reply:
Started this in a very incomplete and incorrect version down below

Reply:
ok I think I've finished it! I tried to make it as simple as possible - e.g., I removed the shell env stuff that I will cover in the `ShellEnv` documentation in the next PR